### PR TITLE
New module git config info

### PR DESCRIPTION
**Pull request by guenhter**
##### SUMMARY

Based on this discussion https://github.com/ansible-collections/community.general/pull/7260#discussion_r1403314182
this  new module for only reading out git configs is now created.

##### ISSUE TYPE

- New Module/Plugin Pull Request

##### COMPONENT NAME

git_config_info

##### ADDITIONAL INFORMATION
The new module is only here to read out git configs and deprecate those functionalities in the `git_config` module

### CODE REVIEW

File: changelogs/fragments/7551-new-git-config-info-module.yml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - git_config_info - Adds a new module for reading git configs.
```
Comment (felixfontein) [inline changelogs/fragments/7551-new-git-config-info-module.yml:2]: Please remove this. New modules must not have changelog fragments.

File: plugins/modules/git_config.py [hunk 2]
```diff
@@ -36,12 +36,14 @@
   list_all:
     description:
       - List all settings (optionally limited to a given O(scope)).
+      - Reading git configs is deprecated and will be removed in community.general 9.0.0. Replacement module is git_config_info.
```
Comment (felixfontein) [inline plugins/modules/git_config.py:39]: Please move the deprecation to a separate PR. Features should only be deprecated a certain time after the feature replacing them has been added, to avoid unnecessary frustration of users who have to support multiple versions of the collection.

(Also removal in 9.0.0 is **way** too early.)

File: plugins/modules/git_config_info.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+---
+module: git_config_info
+author:
+  - Guenther Grill (@guenhter)
+version_added: 8.1.0
+requirements: ['git']
+short_description: Read git configuration
+description:
+  - The M(community.general.git_config_info) module reads the git configuration
+    by invoking 'git config'.
```
Comment (felixfontein) [inline plugins/modules/git_config_info.py:23]:
```suggestion
    by invoking C(git config).
```

File: plugins/modules/git_config_info.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+extends_documentation_fragment:
+  - community.general.attributes
+attributes:
+  check_mode:
+    support: full
+  diff_mode:
+    support: none
```
Comment (felixfontein) [inline plugins/modules/git_config_info.py:30]:
```suggestion
  - community.general.attributes
  - community.general.attributes.info_module
```

File: plugins/modules/git_config_info.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    scope: file
+    path: /etc/gitconfig
+'''
+
+RETURN = '''
+---
+config_value:
+  description: >
+    When O(name) is set, a string containing the value of the setting in name. If O(name) is not set, empty.
+    If a conifg key (e.g. push.pushoption) has more then one entry, just the first one is returned here.
```
Comment (felixfontein) [inline plugins/modules/git_config_info.py:89]:
```suggestion
    If a config key such as V(push.pushoption) has more then one entry, just the first one is returned here.
```

File: plugins/modules/git_config_info.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+  - community.general.attributes
+attributes:
+  check_mode:
+    support: full
+  diff_mode:
+    support: none
+options:
+  name:
+    description:
+      - The name of the setting to read.
```
Comment (felixfontein) [inline plugins/modules/git_config_info.py:34]:
```suggestion
      - The name of the setting to read.
      - If not provided, all settings will be returned as RV(config_values).
```

File: plugins/modules/git_config_info.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    When O(name) is set, a string containing the value of the setting in name. If O(name) is not set, empty.
+    If a conifg key (e.g. push.pushoption) has more then one entry, just the first one is returned here.
+  returned: success
+  type: str
+  sample: "vim"
+
+config_values:
+  description: >
+    When O(name) is not set, a dict containing key/List[value] pairs of all configuration settings.
+    When O(name) is set, a dict containing key/List[value] pairs of the for the defined key.
```
Comment (felixfontein) [inline plugins/modules/git_config_info.py:97]:
```suggestion
  description:
    - This is a dictionary mapping a git configuration setting to a list of its values.
    - When O(name) is not set, all configuration settings are returned here.
    - When O(name) is set, only the setting specified in O(name) is returned here.
      If that setting is not set, the key will still be present, and its value will be an empty list.
```

File: plugins/modules/git_config_info.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    path: /etc/gitconfig
+'''
+
+RETURN = '''
+---
+config_value:
+  description: >
+    When O(name) is set, a string containing the value of the setting in name. If O(name) is not set, empty.
+    If a conifg key (e.g. push.pushoption) has more then one entry, just the first one is returned here.
+  returned: success
```
Comment (felixfontein) [inline plugins/modules/git_config_info.py:90]:
```suggestion
  returned: success if O(name) is set
```

File: plugins/modules/git_config_info.py [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+      - If set to V(file), O(path) must be set to the config file to read from.
+    choices: [ "global", "system", "local", "file" ]
+    default: "system"
+    type: str
+'''
+
+EXAMPLES = '''
+- name: Read a system wide config
+  community.general.git_config_info:
+    name: core.editor
```
Comment (felixfontein) [inline plugins/modules/git_config_info.py:58]:
```suggestion
    name: core.editor
  register: result

- name: Show value of core.editor
  ansible.builtin.debug:
    msg: "{{ result.config_value | default('(not set)', true) }}"
```

File: plugins/modules/git_config_info.py [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            name=dict(type="str"),
+            path=dict(type="path"),
+            scope=dict(required=False, type="str", default="system", choices=["global", "system", "local", "file"]),
+        ),
+        required_if=[
+            ("scope", "local", ["path"]),
+            ("scope", "file", ["path"])
```
Comment (felixfontein) [inline plugins/modules/git_config_info.py:119]:
```suggestion
            ("scope", "file", ["path"]),
```

File: plugins/modules/git_config_info.py [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+    args = [git_path, "config", "--includes", "--" + scope]
+
+    if scope == "file":
+        args.append(path)
+
+    if name:
+        args.extend(["--get-all", name])
+    else:
+        args.append("--list")
+
```
Comment (felixfontein) [inline plugins/modules/git_config_info.py:167]: Why not always append `--null` as well? This makes parsing the result a lot safer (you need to split by `\0` though).

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution!

File: plugins/modules/git_config_info.py (hunk 11 above)
Comment (guenhter) [reply plugins/modules/git_config_info.py:167]: This is a great idea

File: plugins/modules/git_config_info.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+  returned: success if O(name) is set
+  type: str
+  sample: "vim"
+
+config_values:
+  description:
+    - This is a dictionary mapping a git configuration setting to a list of its values.
+    - When O(name) is not set, all configuration settings are returned here.
+    - When O(name) is set, only the setting specified in O(name) is returned here.
+      If that setting is not set, the key will still be present, and its value will be an empty.
```
Comment (felixfontein) [inline plugins/modules/git_config_info.py:101]:
```suggestion
      If that setting is not set, the key will still be present, and its value will be an empty list.
```

File: plugins/modules/git_config_info.py [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+    (rc, out, err) = module.run_command(args, cwd=run_cwd, expand_user_and_vars=False)
+
+    if rc == 128 and "unable to read config file" in err:
+        # This just means nothing has been set at the given scope
+        pass
+    elif rc >= 2:
+        # If the return code is 1, it just means the option hasn't been set yet, which is fine.
+        module.fail_json(rc=rc, msg=err, cmd=" ".join(args))
+
+    output_lines = [line.strip() for line in out.split("\0") if line.strip()]
```
Comment (felixfontein) [inline plugins/modules/git_config_info.py:149]: The `strip()` shouldn't really be necessary anymore with `--null`, so how about:
```suggestion
    output_lines = out.split("\0") if out else []
```
The `if out else []` is necessary to not get `['']` if `out` is empty.

Comment (guenhter) [reply plugins/modules/git_config_info.py]: Thx. This makes a lot of sense. I'm not the python expert as you can see :)

**Review by felixfontein — APPROVED** [review]
Looks good to me. If nobody objects, I'll merge this in a few days.

**Comment by felixfontein** [comment]
@guenhter thanks for your contribution!
